src/H5Dmpio.c

```diff
@@ -4615,26 +5019,69 @@ H5D__mpio_collective_filtered_chunk_reallocate(H5D_filtered_collective_io_info_t
     bool                    need_sort                  = false;
     size_t                  collective_num_entries     = 0;
     size_t                  num_local_chunks_processed = 0;
-    size_t                  i;
-    void                   *gathered_array     = NULL;
-    int                    *counts_disps_array = NULL;
-    int                    *counts_ptr         = NULL;
-    int                    *displacements_ptr  = NULL;
+    void                   *gathered_array             = NULL;
+    int                    *counts_disps_array         = NULL;
+    int                    *counts_ptr                 = NULL;
+    int                    *displacements_ptr          = NULL;
     int                     mpi_code;
     herr_t                  ret_value = SUCCEED;
 
+    typedef struct {
+        H5D_chk_idx_info_t chunk_idx_info;
+        haddr_t            dset_oloc_addr;
+
+        UT_hash_handle hh;
+    } per_dset_info;
+    per_dset_info *dset_info = NULL;
+    per_dset_info *curr_info = NULL;
+    per_dset_info *tmp       = NULL;
+
     FUNC_ENTER_PACKAGE
 
     assert(chunk_list);
     assert(io_info);
-    assert(idx_info);
-    assert(idx_info->storage->idx_type != H5D_CHUNK_IDX_NONE);
+    assert(di);
 
 #ifdef H5Dmpio_DEBUG
     H5D_MPIO_TRACE_ENTER(mpi_rank);
     H5D_MPIO_TIME_START(mpi_rank, "Reallocation of chunk file space");
 #endif
 
+    /* Cache some info about each dataset for later use */
+    for (size_t dset_idx = 0; dset_idx < num_dset_infos; dset_idx++) {
```

Comment:
Would it be possible to build this hash table once and pass it between the different routines?

Reply:
I can probably refactor this for a bit of sharing to avoid the duplicated allocations, but since the code tries to send the minimal amount of data per chunk around through MPI, the hash tables are indexed in two different ways. The `chunk_read` and `chunk_update` functions both use a struct like:

```
    typedef struct {
        const H5D_dset_io_info_t *dset_io_info;
        H5D_fill_buf_info_t       fb_info;
        hsize_t                   file_chunk_size;
        H5S_t                    *fill_space;
        bool                      should_fill;
        bool                      fb_info_init;
        bool                      index_empty;

        UT_hash_handle hh;
    } per_dset_info;
```

and both use the `dset_io_info` field/pointer as a key, so those two functions can share a hash table. The `chunk_reallocate` and `chunk_reinsert` use structs that are close and could be shared if I update the second one to just include the (useless for its case) `dset` pointer:

```
    typedef struct {
        H5D_chk_idx_info_t chunk_idx_info;
        haddr_t            dset_oloc_addr;

        UT_hash_handle hh;
    } per_dset_info;
```

```
    typedef struct {
        H5D_chk_idx_info_t chunk_idx_info;
        haddr_t            dset_oloc_addr;
        H5D_t             *dset;

        UT_hash_handle hh;
    } per_dset_info;
```

Both functions hash based on the `dset_oloc_addr` field, so those two can shared. Let me look into this more, there may be a chance to reduce this down to only two hash table allocations, but maybe just one with some refactoring.

Reply:
@fortnern, I was able to collapse this down so that every function shares the same hash table that is constructed once on setup. However, this revealed some collective metadata issues where some ranks weren't correctly participating in collective metadata reads of the dataset chunk index where they should have been due to the `skip_io` approach in `H5D__read` causing them to skip over certain parts of the I/O code; this issue is somewhat related to #2658.

The most reasonable way to fix this seemed to push an initial collective read of the index into the `io_init` callback in `H5Dchunk.c`. However, this means that even when a dataset is being skipped for I/O, every rank still needs to call the `io_init` and `io_term` callbacks to ensure proper behavior. I could've pushed the logic further up into `H5D__read`/`H5D__write`, but decided not to since it's chunked dataset-specific. If this turns out to be a performance problem, I think it can be mitigated by updating `H5D__chunk_io_init` to skip certain portions of its code if a rank isn't really doing anything.